backend/src/main/java/woowacourse/touroot/authentication/infrastructure/KakaoOauthClient.java

```diff
@@ -0,0 +1,73 @@
+package woowacourse.touroot.authentication.infrastructure;
+
+import java.time.Duration;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.web.client.ClientHttpRequestFactories;
+import org.springframework.boot.web.client.ClientHttpRequestFactorySettings;
+import org.springframework.http.MediaType;
+import org.springframework.http.client.ClientHttpRequestFactory;
+import org.springframework.stereotype.Component;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestClient;
+import woowacourse.touroot.authentication.dto.KakaoAccessTokenResponse;
+import woowacourse.touroot.authentication.dto.OauthUserInformationResponse;
+
+@Component
+public class KakaoOauthClient {
+
+    private final String userInformationRequestUri;
+    private final String accessTokenRequestUri;
+    private final String restApiKey;
+    private final String redirectUri;
+
+    private final RestClient restClient;
+
+    public KakaoOauthClient(
+            @Value("${oauth.kakao.user-information-request-uri}") String userInformationRequestUri,
+            @Value("${oauth.kakao.access-token-request-uri}") String accessTokenRequestUri,
+            @Value("${oauth.kakao.rest-api-key}") String restApiKey,
+            @Value("${oauth.kakao.redirect-uri}") String redirectUri
+    ) {
+
+        ClientHttpRequestFactorySettings settings = ClientHttpRequestFactorySettings.DEFAULTS
+                .withConnectTimeout(Duration.ofSeconds(1))
+                .withReadTimeout(Duration.ofSeconds(3));
+        ClientHttpRequestFactory requestFactory = ClientHttpRequestFactories.get(settings);
+
+        this.userInformationRequestUri = userInformationRequestUri;
+        this.accessTokenRequestUri = accessTokenRequestUri;
+        this.restApiKey = restApiKey;
+        this.redirectUri = redirectUri;
+        this.restClient = RestClient.builder()
+                .requestFactory(requestFactory)
+                .build();
+    }
+
+    public OauthUserInformationResponse requestUserInformation(String authorizationCode) {
+        KakaoAccessTokenResponse kakaoAccessTokenResponse = requestAccessToken(authorizationCode);
+
+        return restClient.get()
+                .uri(userInformationRequestUri)
+                .header("Authorization", "Bearer " + kakaoAccessTokenResponse.accessToken())
+                .retrieve()
+                .toEntity(OauthUserInformationResponse.class)
+                .getBody();
+    }
+
+    private KakaoAccessTokenResponse requestAccessToken(String authorizationCode) {
+        MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
```

Comment:
`MutliValueMap` 사용하신 특별한 이유가 있나요? 지금 코드만 보면 key - value 한 쌍씩 이뤄지는 것 같아서 궁금해용

Reply:
> In cases where our method parameter is of a type MultiValueMap, we can use either the [@RequestParam](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/web/bind/annotation/RequestParam.html) or [@RequestBody](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/web/bind/annotation/RequestBody.html) annotation to bind it appropriately with the body of the HTTP request. That’s because the Servlet API combines the query parameters and form data into a single map called parameters, and that includes automatic parsing of the request body:

html form 데이터가 안정적으로 body에 바인딩 되기 위해서는 Spring에서 util로 제공하는 MultivalueMap 을 이용해야 하는 것으로 알고 있습니다.
관련 링크 첨부드려요!
- [링크](https://www.baeldung.com/spring-url-encoded-form-data)

Reply:
신기하네요